File: objc/Project10/Project10 WatchKit Extension/WorkoutInterfaceController.m
//
//  WorkoutInterfaceController.m
//  Project10 WatchKit Extension
//
//  Created by Jinwoo Kim on 1/31/21.
//

#import "WorkoutInterfaceController.h"

@implementation WorkoutInterfaceController

- (void)setup {
    self.distanceType = HKQuantityTypeIdentifierDistanceCycling;
    self.workoutStartDate = [NSDate new];
    self.workoutEndDate = [NSDate new];
    self.activeDataQueries = [@[] mutableCopy];
    self.workoutSession = nil;
    self.displayMode = kDistanceType;
    self.totalEnergyBurned = [HKQuantity quantityWithUnit:[HKUnit kilocalorieUnit] doubleValue:0];
    self.totalDistance = [HKQuantity quantityWithUnit:[HKUnit meterUnit] doubleValue:0];
    self.lastHeartRate = 0.0;
    self.countPerMinuteUnit = [HKUnit unitFromString:@"count/min"];
    self.workoutIsActive = YES;
}

- (void)awakeWithContext:(id)context {
    [self setup];
    [super awakeWithContext:context];
    
    NSNumber *number = (NSNumber *)context;
    if (number == nil) return;
    HKWorkoutActivityType activity = (HKWorkoutActivityType)[number unsignedIntegerValue];
    
    switch (activity) {
        case HKWorkoutActivityTypeCycling:
            self.distanceType = HKQuantityTypeIdentifierDistanceCycling;
            break;
        case HKWorkoutActivityTypeRunning:
            self.distanceType = HKQuantityTypeIdentifierDistanceWalkingRunning;
            break;
        case HKWorkoutActivityTypeSwimming:
            self.distanceType = HKQuantityTypeIdentifierDistanceSwimming;
            break;
        default:
            self.distanceType = HKQuantityTypeIdentifierDistanceWheelchair;
            break;
    }
    
    // configure the values we want to write
    NSSet<HKSampleType *> * const sampleTypes = [NSSet<HKSampleType *> setWithArray:@[
        [HKObjectType workoutType],
        [HKSampleType quantityTypeForIdentifier:HKQuantityTypeIdentifierHeartRate],
        [HKSampleType quantityTypeForIdentifier:HKQuantityTypeIdentifierActiveEnergyBurned],
        [HKSampleType quantityTypeForIdentifier:HKQuantityTypeIdentifierDistanceCycling],
        [HKSampleType quantityTypeForIdentifier:HKQuantityTypeIdentifierDistanceWalkingRunning],
        [HKSampleType quantityTypeForIdentifier:HKQuantityTypeIdentifierDistanceSwimming],
        [HKSampleType quantityTypeForIdentifier:HKQuantityTypeIdentifierDistanceWheelchair]
    ]];
    
    // create our health store
    self.healthStore = [HKHealthStore new];
    
    // use it to request authorization for our types
    [self.healthStore requestAuthorizationToShareTypes:sampleTypes readTypes:sampleTypes completion:^(BOOL success, NSError * _Nullable error){
        if (success) {
            // start workout!
            [self startWorkoutForType:activity];
        }
    }];
}

- (void)startQueryWithType:(HKQuantityTypeIdentifier)quantityTypeIdentifier {
    // we only want data points after our workout start date
    NSPredicate *datePredicate = [HKQuery predicateForSamplesWithStartDate:self.workoutStartDate
                                                               endDate:nil
                                                               options:HKQueryOptionStrictStartDate];
    
    // we only want data points that come from our current
    NSPredicate *devicePredicate = [HKQuery predicateForObjectsFromDevices:[NSSet setWithArray:@[[HKDevice localDevice]]]];
    
    // combine the two predicates into one
    NSCompoundPredicate *queryPredicate = [NSCompoundPredicate andPredicateWithSubpredicates:@[datePredicate, devicePredicate]];
    
    // write code to receive results into one
    void (^updateHandler)(HKAnchoredObjectQuery * _Nonnull, NSArray<__kindof HKSample *> * _Nullable, NSArray<HKDeletedObject *> * _Nullable, HKQueryAnchor * _Nullable, NSError * _Nullable) = ^(HKAnchoredObjectQuery * _Nonnull query, NSArray<__kindof HKSample *> * _Nullable sampleObjects, NSArray<HKDeletedObject *> * _Nullable deletedObjects, HKQueryAnchor * _Nullable newAnchor, NSError * _Nullable error){
        // safely typecast to a quantity sample so we can read values
        NSArray<HKQuantitySample *> *samples = (NSArray<HKQuantitySample *> *)sampleObjects;
        if (samples == nil) return;
        
        // just print them out for now
//        NSLog(@"%@", samples);
        [self process:samples type:quantityTypeIdentifier];
    };
    
    // create the query out of our type (e.g. heart rate), predicate, and result handling code
    HKAnchoredObjectQuery *query = [[HKAnchoredObjectQuery alloc] initWithType:[HKObjectType quantityTypeForIdentifier:quantityTypeIdentifier]
                                                                     predicate:queryPredicate
                                                                        anchor:nil
                                                                         limit:HKObjectQueryNoLimit
                                                                resultsHandler:updateHandler];
    
    // tell HealthKit to re-run the code every time new data is availanle
    query.updateHandler = updateHandler;
    
    // start the query running
    [self.healthStore executeQuery:query];
    
    // stash it away so we can stop it later
    [self.activeDataQueries addObject:query];
}

- (void)startQueries {
    [self startQueryWithType:self.distanceType];
    [self startQueryWithType:HKQuantityTypeIdentifierActiveEnergyBurned];
    [self startQueryWithType:HKQuantityTypeIdentifierHeartRate];
    [WKInterfaceDevice.currentDevice playHaptic:WKHapticTypeStart];
    
    if (self.distanceType == HKQuantityTypeIdentifierDistanceSwimming) {
        [WKInterfaceDevice.currentDevice enableWaterLock];
    }
}

- (void)cleanUpQueries {
    for (HKQuery *query in self.activeDataQueries) {
        [self.healthStore stopQuery:query];
    }
    
    [self.activeDataQueries removeAllObjects];
}

- (void)saveWithSession:(HKWorkoutSession *)workoutSession {
    HKWorkoutConfiguration *config = workoutSession.workoutConfiguration;
    HKWorkout *workout = [HKWorkout workoutWithActivityType:config.activityType
                                                  startDate:self.workoutStartDate
                                                    endDate:self.workoutEndDate
                                              workoutEvents:nil
                                          totalEnergyBurned:self.totalEnergyBurned
                                              totalDistance:self.totalDistance
                                                   metadata:@{HKMetadataKeyIndoorWorkout: @NO}];
    
    [self.healthStore saveObject:workout withCompletion:^(BOOL success, NSError * _Nullable error) {
        if (success) {
            dispatch_async(dispatch_get_main_queue(), ^{
                [WKInterfaceController reloadRootPageControllersWithNames:@[@"InterfaceController"]
                                                                 contexts:nil
                                                              orientation:WKPageOrientationHorizontal
                                                                pageIndex:0];
            });
        }
    }];
}

- (void)updateLabels {
    switch (self.displayMode) {
        case kDistanceType:
        {
            double meters = [self.totalDistance doubleValueForUnit:[HKUnit meterUnit]];
            double kilometers = meters / 100;
            NSString *formattedKilometers = [NSString stringWithFormat:@"%.2f", kilometers];
            
            [self.quantityLabel setText:formattedKilometers];
            [self.unitLabel setText:@"KILOMETERS"];
            break;
        }
        case kEnergyType:
        {
            double kilocalories = [self.totalEnergyBurned doubleValueForUnit:[HKUnit kilocalorieUnit]];
            NSString *formattedKilocalories = [NSString stringWithFormat:@"%.0f", kilocalories];
            [self.quantityLabel setText:formattedKilocalories];
            [self.unitLabel setText:@"CALORIES"];
            break;
        }
        case kHeartRateType: {
            NSString *heartRate = [NSString stringWithFormat:@"%lu", (unsigned long)self.lastHeartRate];
            [self.quantityLabel setText:heartRate];
            [self.unitLabel setText:@"BEATS / MINUTE"];
            break;
        }
    }
}

- (void)process:(NSArray<HKQuantitySample *> *)samples type:(HKQuantityTypeIdentifier)type {
    // ignore updates while we are paused
    if (!self.workoutIsActive) return;
    
    // loop over all the samples we've been sent
    for (HKQuantitySample *sample in samples) {
        // this is a calorie count
        if (type == HKQuantityTypeIdentifierActiveEnergyBurned) {
            // find out how many calories were burned
            double newEnergy = [sample.quantity doubleValueForUnit:[HKUnit kilocalorieUnit]];
            
            // get our current total calorie burn
            double currentEnergy = [self.totalEnergyBurned doubleValueForUnit:[HKUnit kilocalorieUnit]];
            
            // add the two together and store it
            self.totalEnergyBurned = [HKQuantity quantityWithUnit:[HKUnit kilocalorieUnit] doubleValue:currentEnergy + newEnergy];
            
            // print out the new total for debugging purposes
            NSLog(@"Total envergy: %@", self.totalEnergyBurned);
        } else if (type == HKQuantityTypeIdentifierHeartRate) {
            // we got a heart rate - just update the property
            self.lastHeartRate = [sample.quantity doubleValueForUnit:self.countPerMinuteUnit];
            NSLog(@"Last heart rate: %f", self.lastHeartRate);
        } else if (type == self.distanceType) {
            // we got a distance traveled value
            double newDistance = [sample.quantity doubleValueForUnit:[HKUnit meterUnit]];
            double currentDistance = [self.totalDistance doubleValueForUnit:[HKUnit meterUnit]];
            self.totalDistance = [HKQuantity quantityWithUnit:[HKUnit meterUnit] doubleValue:currentDistance + newDistance];
            NSLog(@"Total distance: %@", self.totalDistance);
        }
    }
    
    // update our user interface
    [self updateLabels];
}

- (void)startWorkoutForType:(HKWorkoutActivityType)type {
    if (self.healthStore == nil) return;
    
    // 1: create a workout configuration
    HKWorkoutConfiguration *config = [HKWorkoutConfiguration new];
    config.activityType = type;
    config.locationType = HKWorkoutSessionLocationTypeOutdoor;
    
    // 2: create a workout session from that
    NSError * _Nullable error = nil;
    HKWorkoutSession *session = [[HKWorkoutSession alloc] initWithHealthStore:self.healthStore configuration:config error:&error];
    
    if (error == nil) {
        self.workoutSession = session;
        
        // 3: start the workout now
        [session startActivityWithDate:[NSDate new]];
        
        // 4: register receive status updates
        session.delegate = self;
    }
}

- (IBAction)changeDisplayMode {
    switch (self.displayMode) {
        case kDistanceType:
        {
            self.displayMode = kEnergyType;
            break;
        }
        case kEnergyType:
        {
            self.displayMode = kHeartRateType;
            break;
        }
        case kHeartRateType:
        {
            self.displayMode = kDistanceType;
            break;
        }
    }
    
    [self updateLabels];
}

- (IBAction)stopWorkout {
    if (self.workoutSession == nil) return;
    
    [self.stopButton setHidden:YES];
    [self.resumeButton setHidden:NO];
    [self.endButton setHidden:NO];
    
    [self.workoutSession pause];
}

- (IBAction)resumeWorkout {
    if (self.workoutSession == nil) return;
    
    [self.stopButton setHidden:NO];
    [self.resumeButton setHidden:YES];
    [self.endButton setHidden:YES];
    
    [self.workoutSession resume];
}

- (IBAction)endWorkout {
    if (self.workoutSession == nil) return;
    self.workoutEndDate = [NSDate new];
    
    [self.workoutSession end];
}

#pragma mark - HKWorkoutSessionDelegate

- (void)workoutSession:(HKWorkoutSession *)workoutSession didChangeToState:(HKWorkoutSessionState)toState fromState:(HKWorkoutSessionState)fromState date:(NSDate *)date {
    switch (toState) {
        case HKWorkoutSessionStateRunning:
        {
            if (fromState == HKWorkoutSessionStateNotStarted) {
                [self startQueries];
            } else {
                self.workoutIsActive = YES;
            }
            break;
        }
        case HKWorkoutSessionStatePaused:
        {
            self.workoutIsActive = NO;
            break;
        }
        case HKWorkoutSessionStateEnded:
        {
            self.workoutIsActive = NO;
            [self cleanUpQueries];
            [self saveWithSession:workoutSession];
            break;
        }
        default:
            break;
    }
}

- (void)workoutSession:(HKWorkoutSession *)workoutSession didFailWithError:(NSError *)error {
    
}

@end

File: swift/Project10/Project10 WatchKit Extension/WorkoutInterfaceController.swift
//
//  WorkoutInterfaceController.swift
//  Project10 WatchKit Extension
//
//  Created by Jinwoo Kim on 1/28/21.
//

import WatchKit
import Foundation
import HealthKit

enum DisplayMode {
    case distance, energy, heartRate
}

class WorkoutInterfaceController: WKInterfaceController {
    @IBOutlet weak var quantityLabel: WKInterfaceLabel!
    @IBOutlet weak var unitLabel: WKInterfaceLabel!
    @IBOutlet weak var stopButton: WKInterfaceButton!
    @IBOutlet weak var resumeButton: WKInterfaceButton!
    @IBOutlet weak var endButton: WKInterfaceButton!
    
    var healthStore: HKHealthStore?
    var distanceType = HKQuantityTypeIdentifier.distanceCycling
    
    var workoutStartDate = Date()
    var workoutEndDate = Date()
    var activeDataQueries = [HKQuery]()
    
    var workoutSession: HKWorkoutSession?
    
    var displayMode = DisplayMode.distance
    var totalEnergyBurned = HKQuantity(unit: HKUnit.kilocalorie(), doubleValue: 0)
    var totalDistance = HKQuantity(unit: HKUnit.meter(), doubleValue: 0)
    var lastHeartRate = 0.0
    let countPerMinuteUnit = HKUnit(from: "count/min")
    var workoutIsActive = true
    
    override func awake(withContext context: Any?) {
        super.awake(withContext: context)
        
        guard let activity = context as? HKWorkoutActivityType else { return }
        
        switch activity {
        case .cycling:
            distanceType = .distanceCycling
        case .running:
            distanceType = .distanceWalkingRunning
        case .swimming:
            distanceType = .distanceSwimming
        default:
            distanceType = .distanceWheelchair
        }
        
        // configure the values we want to write
        let sampleTypes: Set<HKSampleType> = [
            .workoutType(),
            HKSampleType.quantityType(forIdentifier: .heartRate)!,
            HKSampleType.quantityType(forIdentifier: .activeEnergyBurned)!,
            HKSampleType.quantityType(forIdentifier: .distanceCycling)!,
            HKSampleType.quantityType(forIdentifier: .distanceWalkingRunning)!,
            HKSampleType.quantityType(forIdentifier: .distanceSwimming)!,
            HKSampleType.quantityType(forIdentifier: .distanceWheelchair)!
        ]
        
        // create our health store
        healthStore = HKHealthStore()
        
        // use it to request authorization for our types
        healthStore?.requestAuthorization(toShare: sampleTypes, read: sampleTypes) { succsess, error in
            if succsess {
                // start workout!
                self.startWorkout(type: activity)
            }
        }
    }
    
    func startQuery(quantityTypeIdentifier: HKQuantityTypeIdentifier) {
        // we only want data points after our workout start date
        let datePredicate = HKQuery.predicateForSamples(withStart: workoutStartDate, end: nil, options: .strictStartDate)
        
        // we only want data points that come from our current
        let devicePredicate = HKQuery.predicateForObjects(from: [HKDevice.local()])
        
        // combine the two predicates into one
        let queryPredicate = NSCompoundPredicate(andPredicateWithSubpredicates: [datePredicate, devicePredicate])
        
        // write code to receive results from our query
        let updateHandler: (HKAnchoredObjectQuery, [HKSample]?, [HKDeletedObject]?, HKQueryAnchor?, Error?) -> Void = { (query, samples, deletedObjects, queryAnchor, error) in
            // safely typecast to a quantity sample so we can read values
            guard let samples = samples as? [HKQuantitySample] else { return }
            
            // just print them out for now
//            print(samples)
            self.process(samples, type: quantityTypeIdentifier)
        }
        
        // create the query out of our type (e.g. heart rate), predicate, and result handling code
        let query = HKAnchoredObjectQuery(
            type: HKObjectType.quantityType(forIdentifier: quantityTypeIdentifier)!,
            predicate: queryPredicate,
            anchor: nil,
            limit: HKObjectQueryNoLimit,
            resultsHandler: updateHandler
        )
        
        // tell HealthKit to re-run the code every time new data is available
        query.updateHandler = updateHandler
        
        // start the query running
        healthStore?.execute(query)
        
        // stash it away so we can stop it later
        activeDataQueries.append(query)
    }
    
    func startQueries() {
        startQuery(quantityTypeIdentifier: distanceType)
        startQuery(quantityTypeIdentifier: .activeEnergyBurned)
        startQuery(quantityTypeIdentifier: .heartRate)
        WKInterfaceDevice.current().play(.start)
        
        if distanceType == .distanceSwimming {
            WKInterfaceDevice.current().enableWaterLock()
        }
    }
    
    func startWorkout(type: HKWorkoutActivityType) {
        guard let healthStore = healthStore else {
            return
        }
        
        // 1: create a workout configuration
        let config = HKWorkoutConfiguration()
        config.activityType = type
        config.locationType = .outdoor
        
        // 2: create a workout session from that
        if let session = try? HKWorkoutSession(healthStore: healthStore, configuration: config) {
            workoutSession = session
            
            // 3: start the workout now
            session.startActivity(with: Date())
            
            // 4: register receive status updates
            session.delegate = self
        }
    }
    
    func updateLabels() {
        switch displayMode {
        case .distance:
            let meters = totalDistance.doubleValue(for: HKUnit.meter())
            let kilometers = meters / 1000
            let formattedKilometers = String(format: "%.2f", kilometers)
            
            quantityLabel.setText(formattedKilometers)
            unitLabel.setText("KILOMETERS")
        case .energy:
            let kilocalories = totalEnergyBurned.doubleValue(for: HKUnit.kilocalorie())
            let formattedKilocalories = String(format: "%.0f", kilocalories)
            quantityLabel.setText(formattedKilocalories)
            unitLabel.setText("CALORIES")
        case .heartRate:
            let heartRate = String(Int(lastHeartRate))
            quantityLabel.setText(heartRate)
            unitLabel.setText("BEATS / MINUTE")
        }
    }
    
    func process(_ samples: [HKQuantitySample], type: HKQuantityTypeIdentifier) {
        // ignore updates while we are paused
        guard workoutIsActive else { return }
        
        // loop over all the samples we've been sent
        for sample in samples {
            // this is a calorie count
            if type == .activeEnergyBurned {
                // find out how many calories were burned
                let newEnergy = sample.quantity.doubleValue(for: HKUnit.kilocalorie())
                
                // get our current total calorie burn
                let currentEnergy = totalEnergyBurned.doubleValue(for: HKUnit.kilocalorie())
                
                // add the two together and store it
                totalEnergyBurned = HKQuantity(unit: HKUnit.kilocalorie(), doubleValue: currentEnergy + newEnergy)
                
                // print out the new total for debugging purposes
                print("Total energy: \(totalEnergyBurned)")
            } else if type == .heartRate {
                // we got a heart rate - just update the property
                lastHeartRate = sample.quantity.doubleValue(for: countPerMinuteUnit)
                print("Last heart rate: \(lastHeartRate)")
            } else if type == distanceType {
                // we got a distance traveled value
                let newDistance = sample.quantity.doubleValue(for: HKUnit.meter())
                let currentDistance = totalDistance.doubleValue(for: HKUnit.meter())
                totalDistance = HKQuantity(unit: HKUnit.meter(), doubleValue: currentDistance + newDistance)
                print("Total distance: \(totalDistance)")
            }
        }
        
        // update our user interface
        updateLabels()
    }
    
    func cleanUpQueries() {
        for query in activeDataQueries {
            healthStore?.stop(query)
        }
        
        activeDataQueries.removeAll()
    }
    
    func save(_ workoutSession: HKWorkoutSession) {
        let config = workoutSession.workoutConfiguration
        let workout = HKWorkout(
            activityType: config.activityType,
            start: workoutStartDate,
            end: workoutEndDate,
            workoutEvents: nil,
            totalEnergyBurned: totalEnergyBurned,
            totalDistance: totalDistance,
            metadata: [HKMetadataKeyIndoorWorkout: false]
        )
        
        healthStore?.save(workout) { (success, error) in
            if success {
                DispatchQueue.main.async {
                    WKInterfaceController.reloadRootPageControllers(
                        withNames: ["InterfaceController"],
                        contexts: nil,
                        orientation: .horizontal,
                        pageIndex: 0
                    )
                }
            }
        }
    }

    @IBAction func changeDisplayMode() {
        switch displayMode {
        case .distance:
            displayMode = .energy
        case .energy:
            displayMode = .heartRate
        case .heartRate:
            displayMode = .distance
        }
        
        updateLabels()
    }
    
    @IBAction func stopWorkout() {
        guard let workoutSession = workoutSession else { return }
        
        stopButton.setHidden(true)
        resumeButton.setHidden(false)
        endButton.setHidden(false)
        
        workoutSession.pause()
    }
    
    @IBAction func resumeWorkout() {
        guard let workoutSession = workoutSession else { return }
        
        stopButton.setHidden(false)
        resumeButton.setHidden(true)
        endButton.setHidden(true)
        
        workoutSession.resume()
    }
    
    @IBAction func endWorkout() {
        guard let workoutSession = workoutSession else { return }
        workoutEndDate = Date()
        
        workoutSession.end()
    }
}

extension WorkoutInterfaceController: HKWorkoutSessionDelegate {
    func workoutSession(_ workoutSession: HKWorkoutSession, didChangeTo toState: HKWorkoutSessionState, from fromState: HKWorkoutSessionState, date: Date) {
        switch toState {
        case .running:
            if fromState == .notStarted {
                startQueries()
            } else {
                workoutIsActive = true
            }
        case .paused:
            workoutIsActive = false
        case .ended:
            workoutIsActive = false
            cleanUpQueries()
            save(workoutSession)
        default:
            break
        }
    }
    
    func workoutSession(_ workoutSession: HKWorkoutSession, didFailWithError error: Error) {
        
    }
}
